File: src/application/control-server/handlers/workflows.ts
import type { IncomingMessage, ServerResponse } from "node:http";

import {
  buildImportSessionSnapshot,
  importSidecarToGraph,
} from "../../graph-workflow-serializer.js";
import {
  exportAndSaveGraphWorkflow,
  listGraphWorkflows,
  loadGraphWorkflow,
} from "../../graph-workflow-store.js";
import type { GraphWorkflowSaveVariant } from "../../graph-workflow-types.js";
import type { ControlServerDeps } from "../control-server-deps.js";
import { readJsonBody, sendJson } from "../http-utils.js";

function parseSaveVariant(value: unknown): GraphWorkflowSaveVariant | undefined {
  if (value === "playbook" || value === "pipeline" || value === "both") {
    return value;
  }
  return undefined;
}

export async function tryGraphWorkflowRoutes(
  request: IncomingMessage,
  response: ServerResponse,
  deps: ControlServerDeps,
  url: URL,
): Promise<boolean> {
  const { session } = deps;
  const projectRoot = deps.projectRoot ?? process.cwd();

  if (request.method === "GET" && url.pathname === "/api/graph-workflows") {
    sendJson(response, {
      workflows: await listGraphWorkflows(projectRoot),
    });
    return true;
  }
  if (request.method === "POST" && url.pathname === "/api/graph-workflows/export") {
    const body = await readJsonBody(request);
    const workflowId = String(body["workflowId"] ?? "").trim();
    if (!workflowId) {
      sendJson(response, { error: "workflowId is required." }, 400);
      return true;
    }
    const variant = parseSaveVariant(body["variant"]);
    if (!variant) {
      sendJson(response, { error: "variant must be playbook, pipeline, or both." }, 400);
      return true;
    }
    const description = typeof body["description"] === "string" ? body["description"] : undefined;
    const saved = await exportAndSaveGraphWorkflow({
      workflowId,
      description,
      variant,
      graph: session.snapshot().graph,
      projectRoot,
    });
    session.patchGraphWorkflowMetadata({
      linkedWorkflowId: workflowId,
      lastVariant: variant,
      exportedAt: saved.sidecar.updatedAt,
    });
    sendJson(response, saved);
    return true;
  }
  if (request.method === "POST" && url.pathname === "/api/graph-workflows/import") {
    const body = await readJsonBody(request);
    const workflowId = String(body["workflowId"] ?? "").trim();
    if (!workflowId) {
      sendJson(response, { error: "workflowId is required." }, 400);
      return true;
    }
    try {
      const sidecar = await loadGraphWorkflow(workflowId, {
        projectRoot,
      });
      const imported = importSidecarToGraph(sidecar, "playbook");
      session.restoreSnapshot(buildImportSessionSnapshot(imported.graph));
      session.patchGraphWorkflowMetadata({
        linkedWorkflowId: workflowId,
        lastVariant: imported.variant,
      });
      sendJson(response, {
        ...session.snapshot(),
        workflowId,
        importedVariant: imported.variant,
      });
    } catch (error: unknown) {
      const message = error instanceof Error ? error.message : String(error);
      sendJson(response, { error: `Import failed: invalid graph workflow file. ${message}` }, 400);
    }
    return true;
  }
  return false;
}

File: crates/rlm-core/src/control_server/handlers/workflows.rs
use std::sync::Arc;

use axum::extract::State;
use axum::http::StatusCode;
use axum::response::Json;
use serde_json::{json, Value};

use crate::application::graph::{
    build_import_session_snapshot, export_and_save_graph_workflow, import_sidecar_to_graph,
    list_graph_workflows, load_graph_workflow,
};

use super::common::{snapshot_with_extra, ApiError};
use crate::control_server::RouterState;

pub(crate) async fn graph_workflows_list(State(state): State<Arc<RouterState>>) -> Json<Value> {
    let workflows = list_graph_workflows(&state.project_root).unwrap_or_default();
    Json(json!({ "workflows": workflows }))
}

pub(crate) async fn graph_workflows_export(
    State(state): State<Arc<RouterState>>,
    body: Json<Value>,
) -> Result<Json<Value>, ApiError> {
    let workflow_id = body
        .get("workflowId")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim();
    if workflow_id.is_empty() {
        return Err(ApiError {
            status: StatusCode::BAD_REQUEST,
            body: json!({ "error": "workflowId is required." }),
        });
    }
    let variant = body.get("variant").and_then(|v| v.as_str()).unwrap_or("");
    if variant != "playbook" && variant != "pipeline" && variant != "both" {
        return Err(ApiError {
            status: StatusCode::BAD_REQUEST,
            body: json!({ "error": "variant must be playbook, pipeline, or both." }),
        });
    }
    let description = body
        .get("description")
        .and_then(|v| v.as_str())
        .map(String::from);
    let graph = state.session.snapshot().graph;
    let (_path, sidecar) = export_and_save_graph_workflow(
        &state.project_root,
        workflow_id,
        description,
        variant,
        &graph,
    )
    .map_err(|e| ApiError {
        status: StatusCode::BAD_REQUEST,
        body: json!({ "error": e.to_string() }),
    })?;
    state.session.patch_graph_workflow_metadata(
        Some(workflow_id.to_string()),
        Some(variant.to_string()),
        Some(sidecar.updated_at.clone()),
    );
    Ok(Json(
        json!({ "path": format!(".rlm/workflows/{workflow_id}.yaml"), "sidecar": sidecar }),
    ))
}

pub(crate) async fn graph_workflows_import(
    State(state): State<Arc<RouterState>>,
    body: Json<Value>,
) -> Result<Json<Value>, ApiError> {
    let workflow_id = body
        .get("workflowId")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim();
    if workflow_id.is_empty() {
        return Err(ApiError {
            status: StatusCode::BAD_REQUEST,
            body: json!({ "error": "workflowId is required." }),
        });
    }
    let sidecar = load_graph_workflow(&state.project_root, workflow_id).map_err(|e| ApiError {
        status: StatusCode::BAD_REQUEST,
        body: json!({ "error": format!("Import failed: invalid graph workflow file. {e}") }),
    })?;
    let graph = import_sidecar_to_graph(&sidecar, "playbook").map_err(|e| ApiError {
        status: StatusCode::BAD_REQUEST,
        body: json!({ "error": format!("Import failed: invalid graph workflow file. {e}") }),
    })?;
    state
        .session
        .restore_snapshot(build_import_session_snapshot(graph));
    state.session.patch_graph_workflow_metadata(
        Some(workflow_id.to_string()),
        Some("playbook".into()),
        None,
    );
    Ok(Json(snapshot_with_extra(
        &state.session,
        json!({ "workflowId": workflow_id, "importedVariant": "playbook" }),
    )))
}
